hivemind/dht/__init__.py

```diff
@@ -62,16 +90,20 @@ def run(self) -> None:
         uvloop.install()
         loop = asyncio.new_event_loop()
         asyncio.set_event_loop(loop)
-        self.node: DHTNode = loop.run_until_complete(DHTNode.create(
-            initial_peers=list(self.initial_peers), listen_on=self.listen_on, parallel_rpc=self.parallel_rpc,
-            num_workers=self.max_workers or 1, **self.kwargs))
-        self._port.value = self.node.port
-        run_in_background(loop.run_forever)
-        self.ready.set()
+        pipe_awaiter = ThreadPoolExecutor(self.receiver_threads)
+
+        async def _run():
+            node = await DHTNode.create(
+                initial_peers=list(self.initial_peers), listen_on=self.listen_on, parallel_rpc=self.parallel_rpc,
+                num_workers=self.max_workers or 1, **self.kwargs)
+            self._port.value = node.port
+            self.ready.set()
+
+            while True:
+                method, args, kwargs = await loop.run_in_executor(pipe_awaiter, self._pipe.recv)
```

Comment:
JFYI asyncio has methods for [reading](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.connect_read_pipe) from pipes (from my understanding, only raw pipes, so you need to convert to bytes on your own). Not saying this needs to be implemented right away, but we can keep in mind that threads are not necessary here

Reply:
noted
However, in this specific case we only need 1 thread and the only reason we need a thread is to avoid blocking the asyncio tasks while pipe.recv is waiting. So i would argue it is better to use executor than to implement a custom serializer for MPFuture :)